src/Nethermind/Nethermind.Synchronization/SnapSync/SnapServer.cs

```diff
@@ -0,0 +1,239 @@
+                        break;
+                    }
+                    StorageTree sTree = new(_store, storageRoot, _logManager);
+
+                    for (int reqStorage = 1; reqStorage < requestedPath.Length; reqStorage++)
+                    {
+                        byte[]? sRlp = sTree.GetNodeByPath(Nibbles.CompactToHexEncode(requestedPath[reqStorage]));
+                        response.Add(sRlp);
+                    }
+                    break;
+            }
+        }
+
+        return response.ToArray();
+    }
+
+    public byte[][] GetByteCodes(Keccak[] requestedHashes, long byteLimit)
+    {
+        _logger.Info("GetByteCodes");
+        long currentByteCount = 0;
+        List<byte[]> response = new();
+
+        if (byteLimit > HardResponseByteLimit)
+        {
+            byteLimit = HardResponseByteLimit;
+        }
+
+        foreach (Keccak codeHash in requestedHashes)
+        {
+            // break when the response size exceeds the byteLimit - it is a soft limit
+            // so not a big issue if we so over slightly.
+            if (currentByteCount > byteLimit)
+            {
+                break;
+            }
+
+            // TODO: handle empty code in the DB itself?
+            if (codeHash.Bytes.SequenceEqual(Keccak.OfAnEmptyString.Bytes))
+            {
+                response.Add(new byte[] { });
+            }
+            byte[]? code = _dbProvider.CodeDb.Get(codeHash);
+            if (code is null) continue;
+            response.Add(code);
+            currentByteCount += code.Length;
+        }
+
+        return response.ToArray();
+    }
+
+    public (PathWithAccount[], byte[][]) GetAccountRanges(Keccak rootHash, Keccak startingHash, Keccak? limitHash, long byteLimit)
+    {
+        _logger.Info("GetAccountRanges");
+
+        (Dictionary<byte[], byte[]>? requiredNodes, long _, bool _) = GetNodesFromTrieVisitor(rootHash, startingHash,
+            limitHash ?? Keccak.MaxValue, byteLimit, HardResponseByteLimit, isStorage: false);
+        StateTree tree = new(_store, _logManager);
+
+        PathWithAccount[] nodes = new PathWithAccount[requiredNodes.Count];
+        int index = 0;
+        foreach (PathWithAccount? result in requiredNodes.Select(res => new PathWithAccount(new Keccak(res.Key), _decoder.Decode(new RlpStream(res.Value)))))
+        {
+            nodes[index] = result;
+            index += 1;
+        }
+
+        if (nodes.Length == 0) return (nodes, Array.Empty<byte[]>());
+        byte[][]? proofs = GenerateRangeProof(tree, startingHash, nodes[^1].Path, rootHash);
+        return (nodes, proofs);
+    }
+
+    public (PathWithStorageSlot[][], byte[][]?) GetStorageRanges(Keccak rootHash, PathWithAccount[] accounts, Keccak? startingHash, Keccak? limitHash, long byteLimit)
+    {
+        _logger.Info("GetStorageRanges");
+        long responseSize = 0;
+        StateTree tree = new(_store, _logManager);
+        List<PathWithStorageSlot[]> responseNodes = new();
```

Comment:
As I see use of List here two questions: (for whole file)
1. Can we pre-allocate them to desired size to avoid resizing?
2. Can we change the results from arrays to List so we avoid ToArray? We can and sometimes use different types in network messages to avoid heavy allocations like in `HashesMessage` we use `IReadOnlyList`

Reply:
- initialized the List with max possible capacity ( there are some cases where we cant know the size, but we have a upper limit)
 - Changed List<byte[]> to IList<byte[]>, so that Message can accept both byte[][] and List<byte[]>